tree/ntuple/v7/src/RClusterPool.cxx

```diff
@@ -0,0 +1,329 @@
+         bool discard = false;
+         {
+            std::unique_lock<std::mutex> lock(fLockWorkQueue);
+            for (auto &inFlight : fInFlightClusters) {
+               if (inFlight.fClusterId != item.fClusterId)
+                  continue;
+               discard = inFlight.fIsExpired;
+               break;
+            }
+         }
+         if (discard)
+            cluster.reset();
+
+         item.fPromise.set_value(std::move(cluster));
+      }
+   } // while (true)
+}
+
+ROOT::Experimental::Detail::RCluster *
+ROOT::Experimental::Detail::RClusterPool::FindInPool(DescriptorId_t clusterId) const
+{
+   for (const auto &cptr : fPool) {
+      if (cptr && (cptr->GetId() == clusterId))
+         return cptr.get();
+   }
+   return nullptr;
+}
+
+size_t ROOT::Experimental::Detail::RClusterPool::FindFreeSlot() const
+{
+   auto N = fPool.size();
+   for (unsigned i = 0; i < N; ++i) {
+      if (!fPool[i])
+         return i;
+   }
+
+   R__ASSERT(false);
+   return N;
+}
+
+
+namespace {
+
+/// Helper class for the (cluster, column list) pairs that should be loaded in the background
+class RProvides {
+   using DescriptorId_t = ROOT::Experimental::DescriptorId_t;
+   using ColumnSet_t = ROOT::Experimental::Detail::RPageSource::ColumnSet_t;
+
+private:
+   std::map<DescriptorId_t, ColumnSet_t> fMap;
+
+public:
+   void Insert(DescriptorId_t clusterId, const ColumnSet_t &columns)
+   {
+      fMap.emplace(clusterId, columns);
+   }
+
+   bool Contains(DescriptorId_t clusterId) {
+      return fMap.count(clusterId) > 0;
+   }
+
+   void Erase(DescriptorId_t clusterId, const ColumnSet_t &columns)
+   {
+      auto itr = fMap.find(clusterId);
+      if (itr == fMap.end())
+         return;
+      ColumnSet_t d;
+      std::copy_if(itr->second.begin(), itr->second.end(), std::inserter(d, d.end()),
+         [&columns] (DescriptorId_t needle) { return columns.count(needle) == 0; });
+      if (d.empty()) {
+         fMap.erase(itr);
+      } else {
+         itr->second = d;
+      }
+   }
+
+   decltype(fMap)::iterator begin() { return fMap.begin(); }
+   decltype(fMap)::iterator end() { return fMap.end(); }
+};
+
+} // anonymous namespace
+
+ROOT::Experimental::Detail::RCluster *
+ROOT::Experimental::Detail::RClusterPool::GetCluster(
+   DescriptorId_t clusterId, const RPageSource::ColumnSet_t &columns)
+{
+   const auto &desc = fPageSource.GetDescriptor();
+
+   // Determine previous cluster ids that we keep if they happen to be in the pool
+   std::set<DescriptorId_t> keep;
+   auto prev = clusterId;
+   for (unsigned int i = 0; i < fWindowPre; ++i) {
+      prev = desc.FindPrevClusterId(prev);
+      if (prev == kInvalidDescriptorId)
+         break;
+      keep.insert(prev);
+   }
+
+   // Determine following cluster ids and the column ids that we want to make available
+   RProvides provide;
+   provide.Insert(clusterId, columns);
+   auto next = clusterId;
+   for (unsigned int i = 1; i < fWindowPost; ++i) {
```

Comment:
Are the cluster always guaranteed to be the same size?  (For example, they might not be in case of merging 2 RNtuple that were created differently).  If the clusters can not be guaranteed to be the same size then the memory cost of "N" clusters in memory (I assume that is the meaning of fWindowPost) can vary (possibly widely).  So it may (or may not) be easier for the user to configure in term of memory size.  

Another source of cluster size being different is when fetching just a sub-set of columns. (where the aggregate of all columns may have a consistent size for a cluster but not every sub-sets).

Reply:
I agree, clusters don't have the same size and we eventually want the memory consumption be the criterion. I added a TODO:

    // TODO(jblomer): instead of a fixed-sized window, eventually we should determine the window size based on
    // a user-defined memory limit.  The size of the preloaded data can be determined at the beginning of
    // GetCluster from the descriptor and the current contents of fPool.